introspection-engine/connectors/mongodb-introspection-connector/src/sampler/statistics.rs

```diff
@@ -485,9 +486,86 @@ fn populate_fields(
         warnings.push(crate::warnings::fields_with_empty_names(&fields_with_empty_names));
     }
 
+    filter_out_empty_types(&mut models, &mut types, warnings);
+
     (models, types)
 }
 
+/// From the resulting data model, remove all types with no fields and change
+/// the field types to Json.
+fn filter_out_empty_types(
+    models: &mut BTreeMap<String, Model>,
+    types: &mut BTreeMap<String, CompositeType>,
+    warnings: &mut Vec<Warning>,
+) {
+    let mut fields_with_an_empty_type = Vec::new();
+
+    let empty_types: HashSet<String> = types
+        .iter()
+        .filter_map(|(name, r#type)| {
+            if r#type.fields.is_empty() {
+                Some(name.clone())
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    static DOCS: &str = "Nested objects had no data in the sample dataset to introspect a nested type.";
+
+    // 1. remove all types that have no fields.
+    types.retain(|_, v| !v.fields.is_empty());
+
+    // 2. change all fields in models that point to a non-existing type to Json.
+    for (model_name, model) in models.iter_mut() {
+        let mut keep = VecDeque::with_capacity(model.fields.len());
+
+        while let Some(mut field) = model.fields.pop() {
+            if let Field::ScalarField(sf) = &mut field {
+                match &sf.field_type {
+                    datamodel::FieldType::CompositeType(ct) if !types.contains_key(ct) => {
+                        fields_with_an_empty_type.push((Name::Model(model_name.clone()), sf.name.clone()));
+                        sf.field_type = datamodel::FieldType::Scalar(datamodel::ScalarType::Json, None, None);
+                        sf.documentation = Some(DOCS.to_owned());
+                    }
+                    _ => (),
+                }
+            };
+
+            keep.push_front(field);
+        }
```

Comment:
I am wondering why we need to move things here, wouldn't mutating the field's type in place be simpler?

Reply:
We did need to move. Not anymore in the final version.